xdsl/irdl/constraints.py

```diff
@@ -20,38 +21,35 @@ class ConstraintContext:
     Contains the assignment of constraint variables.
     """
 
-    _variables: dict[str, Attribute] = field(default_factory=dict)
+    _variables: ScopedDict[str, Attribute] = field(default_factory=lambda: ScopedDict())
     """The assignment of constraint variables."""
 
-    _range_variables: dict[str, tuple[Attribute, ...]] = field(default_factory=dict)
+    _range_variables: ScopedDict[str, tuple[Attribute, ...]] = field(
+        default_factory=lambda: ScopedDict()
+    )
     """The assignment of constraint range variables."""
 
-    def get_variable(self, key: str) -> Attribute:
-        return self._variables[key]
+    def get_variable(self, key: str) -> Attribute | None:
+        return self._variables.get(key)
 
-    def get_range_variable(self, key: str) -> tuple[Attribute, ...]:
-        return self._range_variables[key]
+    def get_range_variable(self, key: str) -> tuple[Attribute, ...] | None:
+        return self._range_variables.get(key)
 
     def set_variable(self, key: str, attr: Attribute):
         self._variables[key] = attr
 
     def set_range_variable(self, key: str, attr: tuple[Attribute, ...]):
         self._range_variables[key] = attr
 
-    @property
-    def variables(self) -> Sequence[str]:
-        return tuple(self._variables.keys())
+    def push_variable_scope(self, name: str | None = None):
+        self._variables = ScopedDict(self._variables, name=name)
+        self._range_variables = ScopedDict(self._range_variables, name=name)
 
-    @property
-    def range_variables(self) -> Sequence[str]:
-        return tuple(self._range_variables.keys())
```

Comment:
Are these just dead code? Great that we're removing them if so

Reply:
they are after my change, it feels quite silly that we were constructing and throwing away the tuples so many times. I am very much looking forward to ASV being set up properly to track the impact of these sorts of changes